### PR TITLE
Ddcnls 6979

### PR DESCRIPTION
**Pull request by thirlz95**
- [x] Unit tests passing
- [x] Coverage reached
- [x] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: app/views/page_not_found.scala.html [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+<a href="@url">@serviceName</a>
+}
+
+@main(
+    title = Messages("tai.page_not_found.tabtitle")
+){
+<div class="grid-row">
+    <div class="content">
+
+        <div class="row">
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: class row isn't doing much so it can go

Comment (Muhammad-Ali-Ahmad) [reply app/views/page_not_found.scala.html]: instead make it column-two-thirds that wraps both heading and content

File: app/views/page_not_found.scala.html [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            <h1 id="header" class="heading-xlarge">@Messages("tai.page_not_found.heading")</h1>
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: don't need id or class heading-xlarge since we haven't used extra large heading on any of the other services either

File: app/views/page_not_found.scala.html [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * limitations under the License.
+ *@
+
+@import play.api.Play.current
+@import uk.gov.hmrc.play.config.RunMode._
+@import uk.gov.hmrc.tai.config.ApplicationConfig
+
+@import layouts.{govuk_template => hmrcGovUkTemplate}
+@import uk.gov.hmrc.play.views.html.{layouts => uiLayouts}
+@import play.api.i18n.Messages
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: looks like you don't actually need any of these imports. It just works without it - even without the messages since it is included implicitly

File: app/views/page_not_found.scala.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@import play.api.Play.current
+@import uk.gov.hmrc.play.config.RunMode._
+@import uk.gov.hmrc.tai.config.ApplicationConfig
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: Dont need these imports either so they can go

File: app/views/page_not_found.scala.html [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+@anchor(serviceName: String, url: String) = {
+<a href="@url">@serviceName</a>
+}
+
+@main(
+    title = Messages("tai.page_not_found.tabtitle")
+){
+
+<div class="grid-row">
+    <div class="content">
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: Dont need this div or class

File: app/views/page_not_found.scala.html [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        <div class="column-two-thirds">
+            <h1>@Messages("tai.page_not_found.heading")</h1>
+        </div>
+
+        <div class="column-two-thirds">
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/page_not_found.scala.html]: Dont need second div with class column-two-thirds. One div with class column-two-thirds can wrap h1 and paragraphs

File: conf/messages [hunk 7]
```diff
@@ -1459,6 +1459,16 @@ tai.updateIncome.CYPlus1.same.paragraph1=You have entered {0}
 tai.updateIncome.CYPlus1.same.paragraph2=This is the same as the amount already estimated for {0} for next year, so you do not need to update your income.
 tai.updateIncome.CYPlus1.same.cancel=Return to your Income Tax Summary
 
+tai.page_not_found.tabtitle = Page not found - Check your Income Tax - GOV.UK
```
Comment (Muhammad-Ali-Ahmad) [inline conf/messages]: You don't need a separate tab title with "- Check your Income Tax - GOV.UK". You can just have tai.page_not_found.heading and that can be used for both title and h1. The main template already adds service title and GOV.UK at the end of the page title - If you look at the browser title now, it says "Page not found - Check your Income Tax - GOV.UK - Check your Income Tax - GOV.UK"

File: app/views/page_not_found.scala.html (hunk 4 above)
Comment (thirlz95) [reply app/views/page_not_found.scala.html]: Cool, I've removed them

File: app/views/page_not_found.scala.html (hunk 5 above)
Comment (thirlz95) [reply app/views/page_not_found.scala.html]: removed

File: app/views/page_not_found.scala.html (hunk 6 above)
Comment (thirlz95) [reply app/views/page_not_found.scala.html]: removed it being declared twice, wrapped it around the parent div like so: <div class="content column-two-thirds">

File: conf/messages (hunk 7 above)
Comment (thirlz95) [reply conf/messages]: cheers for the explanation regarding this, it had confused me when it printed Page not found - Check your Income Tax - GOV.UK - Check your Income Tax - GOV.UK", I had not released the 'Check your Income Tax - GOV.UK' would be appended to within the . main template for the heading field. In future I'll look out for this.

File: conf/messages [hunk 8]
```diff
@@ -1459,6 +1459,16 @@ tai.updateIncome.CYPlus1.same.paragraph1=You have entered {0}
 tai.updateIncome.CYPlus1.same.paragraph2=This is the same as the amount already estimated for {0} for next year, so you do not need to update your income.
 tai.updateIncome.CYPlus1.same.cancel=Return to your Income Tax Summary
 
+tai.page_not_found.tabtitle = Page not found
```
Comment (Muhammad-Ali-Ahmad) [inline conf/messages]: tabtitle and heading is the same. please remove one and do same for welsh messages file

File: app/views/main.scala.html [hunk 9]
```diff
@@ -128,6 +132,8 @@
 
 @inlineScript = {
   <script>
+
+    console.log()
```
Comment (Muhammad-Ali-Ahmad) [inline app/views/main.scala.html]: remove console log

**Comment by danewatts** [comment]
@thirlz95 I cant review code I have helped write. Ask in our team channel for review

File: app/controllers/IncomeController.scala [hunk 10]
```diff
@@ -19,7 +19,7 @@ package controllers
 import com.google.inject.name.Named
 import javax.inject.Inject
 import controllers.actions.ValidatePerson
-import controllers.auth.{AuthAction, AuthedUser}
+import controllers.auth.{AuthAction, AuthedUser, AuthenticatedRequest}
```
Comment (frikit) [inline app/controllers/IncomeController.scala]: Look like this is not used, please delete it

Comment (thirlz95) [reply app/controllers/IncomeController.scala]: removed AuthenticatedRequest import

File: conf/application.conf [hunk 11]
```diff
@@ -318,6 +318,10 @@ govuk-tax {
         host = localhost
         port = 9331
       }
+      auth {
+        host = localhost
+        port = 9331
+      }
```
Comment (oc-ste) [inline conf/application.conf]: Why was this added?

File: app/views/main.scala.html [hunk 12]
```diff
@@ -40,7 +40,8 @@
   gaCustomTitle: Option[String] = None,
   optimizelyAudience: Option[String] = None,
   pageHeader: Option[Html] = None,
-  script: Option[Html] = None
+  script: Option[Html] = None,
+  isAuthenticated: Boolean = true
```
Comment (oc-ste) [inline app/views/main.scala.html]: defaulting isAuthenticated is confusing as it suggests every page in tai-frontend is authenticated. We should be asking auth if it's authenticated and not setting it ourselves.

File: test/controllers/FakeTaiPlayApplication.scala [hunk 13]
```diff
@@ -44,10 +47,20 @@ trait FakeTaiPlayApplication extends OneServerPerSuite with PatienceConfiguratio
...
+    "microservice.metrics.graphite.enabled"                 -> false,
+    "metrics.jvm"                                           -> false
   )
 
-  implicit override lazy val app: Application = new GuiceApplicationBuilder().configure(additionalConfiguration).build()
+  implicit override lazy val app: Application = new GuiceApplicationBuilder()
+    .configure(additionalConfiguration)
+    .overrides(
+      bind[Metrics].toInstance(mock[Metrics]),
+      bind[MetricsFilter].toInstance(mock[MetricsFilter])
```
Comment (oc-ste) [inline test/controllers/FakeTaiPlayApplication.scala]: Why are these two added? No tests have changed

File: app/views/main.scala.html [hunk 14]
```diff
@@ -77,8 +78,11 @@
 }
 
 @signOutUrl = @{
-    mustacheCheck {
-        routes.ServiceController.serviceSignout().url
+    mustacheCheck{
+        val signOutLink = routes.ServiceController.serviceSignout().url
+        authedUser.fold{
+            if(isAuthenticated) signOutLink else ""
```
Comment (oc-ste) [inline app/views/main.scala.html]: If there is an authedUser, it's already authenticated, you don't need this boolean

File: app/uk/gov/hmrc/tai/config/ApplicationGlobal.scala [hunk 15]
```diff
@@ -63,19 +70,21 @@ class TaiErrorHandler @Inject()(
...
-    )
-  )
+  override def onClientError(request: RequestHeader, statusCode: Int, message: String): Future[Result] = {
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+    implicit val req: Request[_] = Request(request, "")
+    statusCode match {
+      case play.mvc.Http.Status.NOT_FOUND =>
+        authorised() {
+          Future.successful(NotFound(views.html.page_not_found(isAuthenticated = true)))
```
Comment (oc-ste) [inline app/uk/gov/hmrc/tai/config/ApplicationGlobal.scala]: It's better to have two views here rather than pass in the bool for two different views

**Comment by oc-ste** [comment]
We're redone this work on another branch
